Sources/BuildServerIntegration/BuildServerManager.swift

```diff
@@ -670,31 +668,76 @@ package actor BuildServerManager: QueueBasedMessageHandler {
-        return true
+    await self.buildTargetsDidChange(.didChangeBuildTargets(changedTargets: changedTargets))
+  }
+
+  private enum BuildTargetsChange {
+    case didChangeBuildTargets(changedTargets: Set<BuildTargetIdentifier>?)
+    case buildTargetsReceivedResultAfterTimeout(
+      request: WorkspaceBuildTargetsRequest,
+      newResult: [BuildTargetIdentifier: BuildTargetInfo]
+    )
+    case sourceFilesReceivedResultAfterTimeout(
+      request: BuildTargetSourcesRequest,
+      newResult: BuildTargetSourcesResponse
+    )
+  }
+
+  /// Update the cached state in `BuildServerManager` because new data was received from the BSP server.
+  ///
+  /// This handles a few seemingly unrelated reasons to ensure that we think about which caches to invalidate in the
+  /// other scenarios as well, when making changes in here.
+  private func buildTargetsDidChange(_ stateChange: BuildTargetsChange) async {
+    let changedTargets: Set<BuildTargetIdentifier>?
+
+    switch stateChange {
+    case .didChangeBuildTargets(let changedTargetsValue):
+      changedTargets = changedTargetsValue
+      self.cachedAdjustedSourceKitOptions.clear(isolation: self) { cacheKey in
+        guard let changedTargets else {
+          // All targets might have changed
+          return true
+        }
+        return changedTargets.contains(cacheKey.target)
       }
-      return updatedTargets.contains(cacheKey.target)
-    }
-    self.cachedBuildTargets.clearAll(isolation: self)
-    self.cachedTargetSources.clear(isolation: self) { cacheKey in
-      guard let updatedTargets else {
-        // All targets might have changed
-        return true
+      self.cachedBuildTargets.clearAll(isolation: self)
+      self.cachedTargetSources.clear(isolation: self) { cacheKey in
+        guard let changedTargets else {
+          // All targets might have changed
+          return true
+        }
+        return !changedTargets.intersection(cacheKey.targets).isEmpty
       }
-      return !updatedTargets.intersection(cacheKey.targets).isEmpty
-    }
+    case .buildTargetsReceivedResultAfterTimeout(let request, let newResult):
+      changedTargets = nil
+
+      // Caches not invalidated:
+      // - cachedAdjustedSourceKitOptions: We would not have requested SourceKit options for targets that we didn't
+      //   know about. Even if we did, the build server now telling us about the target should not change the options of
+      //   the file within the target
+      // - cachedTargetSources: Similar to cachedAdjustedSourceKitOptions, we would not have requested sources for
+      //   targets that we didn't know about and if we did, they wouldn't be affected
+      self.cachedBuildTargets.set(request, to: newResult)
+    case .sourceFilesReceivedResultAfterTimeout(let request, let newResult):
+      changedTargets = Set(request.targets)
+
+      // Caches not invalidated:
+      // - cachedAdjustedSourceKitOptions: Same as for buildTargetsReceivedResultAfterTimeout.
```

Comment:
Assuming that the files are different, we probably do need to clear options here? ie. there may now be added/removed files and thus we need to re-request options.

I also wonder if this is true for build targets - that they have changed could imply that new search paths are needed (or old removed). For SwiftPM this isn't relevant today (as it has the one folder that everything is built into), but for Bazel I imagine this is much more relevant?

And lastly, is it now possible that we have multiple target/source requests in flight in different orders?